Include support for openid

chrisjuchem opened the pull request:
My patch discussed in [this issue](https://github.com/BitySA/oauth2-auth-code-pkce/issues/5) has now been stable for a few weeks and has been working well.

This adds a config option to explicitly expose any tokens besides the access token (namely, the `id_token`) and config options for extra parameters to pass to the authorize and refresh routes (In my case, auth0 needed the `client_id` for the refresh route, and I needed to be able to specify an `audience` during the initial authorization).

chrisjuchem commented on index.ts:
```diff
@@ -390,10 +423,22 @@ export class OAuth2AuthCodePKCE {
+        scopes
+      };
     })
-    .catch(jsonPromise => Promise.reject(jsonPromise))
```
This appears redundant to me, let me know if I'm mistaken.

chrisjuchem commented on index.ts:
```diff
@@ -402,7 +447,7 @@ export class OAuth2AuthCodePKCE {
           break;
       }
-      return Promise.reject(error);
+      return Promise.reject(toErrorClass(error));
```
This may be a breaking change depending on usage, but I thought it should be consistent with other errors.

lf94 replied:
No this is probably the correct thing to do - thanks!

lf94 commented:
Thanks @chrisjuchem , I'll look over this in the coming days, been very busy!

danielmcormond commented:
I would use this too, looks like `extraRefreshParams` would allow me to send some additional fields that IdentityServer4 expects.

lf94 commented on index.ts:
```diff
@@ -298,6 +304,13 @@ export class OAuth2AuthCodePKCE {
+              builtUrl+`&${key}=${encodeURIComponent(val)}`,
+          url)
+    }
+
```
Line 376 has the same process. Please create a function with abstracts this process and to prevent accidental mis-behaviors.

```
url = `${url}&${dictToQueryString(extraAuthorizeParams)}`

function dictToQueryString(dict: {[k: string]: string}): string { ... }
```

lf94 commented on index.ts:
```diff
@@ -382,6 +406,15 @@ export class OAuth2AuthCodePKCE {
+          (a: ObjStringDict, token: string) =>
+            json[token] ? { ...a, [token]: json[token] } : a,
+          {}
+        );
```
I think this code could more easily be understood using .map().filter(). Could you please change it to use those? While *I* understand it, I can easily see someone wanting to glance over this .reduce().

Like so:

```
tokensToExpose = Object.fromEntries(explicitlyExposedTokens
  .map((token: string) => [token, json[token]])
  .filter((property: object | undefined) => property !== undefined));
```

I think it's clear to see this is much easier to understand. :slightly_smiling_face: 

(Also I realize you used .reduce() to transform the array into a dictionary/object, but there's https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/fromEntries which can easily handle this.)

lf94 commented on index.ts:
```diff
@@ -527,6 +575,15 @@ export class OAuth2AuthCodePKCE {
+            (a: ObjStringDict, token: string) =>
+              json[token] ? { ...a, [token]: json[token] } : a,
+            {}
+          );
```
Same as above, use map+filter.

lf94 commented on index.ts:
```diff
@@ -390,10 +423,22 @@ export class OAuth2AuthCodePKCE {
+      return {
+        token: accessToken,
+        scopes
+      };
```
There's no need to repeat the same structure twice: could you please change this so it just adds the property to the existing object?

Such as:
```
let accessContext = { token, scopes };
if (Object.keys(tokensToExpose).length > 0) { accessContext.explicitlyExposedTokens = tokensToExpose; }
return accessContext;
```
It reduces a lot of noise :slightly_smiling_face:

lf94 commented on index.ts:
```diff
@@ -535,7 +592,19 @@ export class OAuth2AuthCodePKCE {
+        return {
+          token: accessToken,
+          scopes
+        };
```
Same as mentioned above: please reduce this to just adding the property.

chrisjuchem commented:
@lf94 Thanks for the feedback. I've addressed your comments and also added an extra option to pass params in for a single authorize call, as that became important for our use case.

lf94 commented:
Excellent @chrisjuchem , things look good. I'll test it out tomorrow! Out of curiosity, are you able to say who "our" is? :slightly_smiling_face:

murm commented:
Hi, it would be really awesome if this was available very soon. We are using this package and today we actually have a need to change which token we get.

Any idea when it will be merged? :)

best regards
murm

lf94 commented:
@chrisjuchem @murm this is now deployed as 2.10.1! :slightly_smiling_face:
